src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDependency.cs

```diff
@@ -0,0 +1,1057 @@
+
+            internal void Invoke(SqlNotificationEventArgs args)
+            {
+                _args = args;
+                ExecutionContext.Run(_context, s_contextCallback, this);
+            }
+
+            private static void InvokeCallback(object eventContextPair)
+            {
+                EventContextPair pair = (EventContextPair)eventContextPair;
+                pair._eventHandler(pair._dependency, (SqlNotificationEventArgs)pair._args);
+            }
+        }
+
+        // Instance members
+
+        // SqlNotificationRequest required state members
+
+        // Only used for SqlDependency.Id.
+        private readonly string _id = Guid.NewGuid().ToString() + ";" + s_appDomainKey;
+        private string _options; // Concat of service & db, in the form "service=x;local database=y".
+        private int _timeout;
+
+        // Various SqlDependency required members
+        private bool _dependencyFired = false;
+        // We are required to implement our own event collection to preserve ExecutionContext on callback.
+        private List<EventContextPair> _eventList = new List<EventContextPair>();
+        private object _eventHandlerLock = new object(); // Lock for event serialization.
+        // Track the time that this dependency should time out. If the server didn't send a change
+        // notification or a time-out before this point then the client will perform a client-side 
+        // timeout.
+        private DateTime _expirationTime = DateTime.MaxValue;
+        // Used for invalidation of dependencies based on which servers they rely upon.
+        // It's possible we will over invalidate if unexpected server failure occurs (but not server down).
+        private List<string> _serverList = new List<string>();
+
+        // Static members
+
+        private static object s_startStopLock = new object();
+        private static readonly string s_appDomainKey = Guid.NewGuid().ToString();
+        // Hashtable containing all information to match from a server, user, database triple to the service started for that 
+        // triple.  For each server, there can be N users.  For each user, there can be N databases.  For each server, user, 
+        // database, there can only be one service.
+        private static Dictionary<string, Dictionary<IdentityUserNamePair, List<DatabaseServicePair>>> s_serverUserHash =
+                   new Dictionary<string, Dictionary<IdentityUserNamePair, List<DatabaseServicePair>>>(StringComparer.OrdinalIgnoreCase);
+        private static SqlDependencyProcessDispatcher s_processDispatcher = null;
+        // The following two strings are used for AppDomain.CreateInstance.
+        private static readonly string s_assemblyName = (typeof(SqlDependencyProcessDispatcher)).Assembly.FullName;
+        private static readonly string s_typeName = (typeof(SqlDependencyProcessDispatcher)).FullName;
+
+        // Constructors
+
+        [HostProtection(ExternalThreading = true)]
```

Comment:
I believe `HostProtection` can be deleted. See https://github.com/dotnet/corefx/pull/12081#discussion_r80806580 and https://github.com/dotnet/coreclr/pull/8610